src/shared_modules/content_manager/src/components/offlineDownloader.hpp

```diff
@@ -0,0 +1,158 @@
+ * of responsibility.
+ *
+ */
+class OfflineDownloader final : public AbstractHandler<std::shared_ptr<UpdaterContext>>
+{
+private:
+    /**
+     * @brief Pushes the state of the current stage into the data field of the context.
+     *
+     * @param contextData Reference to the context data.
+     * @param status Status to be pushed.
+     */
+    void pushStageStatus(nlohmann::json& contextData, std::string status) const
+    {
+        auto statusObject = nlohmann::json::object();
+        statusObject["stage"] = "OfflineDownloader";
+        statusObject["status"] = std::move(status);
+
+        contextData.at("stageStatus").push_back(std::move(statusObject));
+    }
+
+    /**
+     * @brief Function to calculate the hash of a file.
+     *
+     * @param filepath Path to the file.
+     * @return std::string Digest vector.
+     */
+    std::string hashFile(const std::filesystem::path& filepath) const
+    {
+        if (std::ifstream inputFile(filepath, std::fstream::in); inputFile)
+        {
+            constexpr int BUFFER_SIZE {4096};
+            std::array<char, BUFFER_SIZE> buffer;
+
+            Utils::HashData hash;
+            while (inputFile.read(buffer.data(), buffer.size()))
+            {
+                // LCOV_EXCL_START
+                hash.update(buffer.data(), inputFile.gcount());
+                // LCOV_EXCL_STOP
+            }
+            hash.update(buffer.data(), inputFile.gcount());
+
+            return Utils::asciiToHex(hash.hash());
+        }
+
+        throw std::runtime_error {"Unable to open '" + filepath.string() + "' for hashing."};
+    };
+
+    /**
+     * @brief Downloads the requested local file and update the context accordingly.
+     *
+     * @note Despite the method name, there is no such download since the file is present in the filesystem.
+     *
+     * @param context Updater context.
+     */
+    void download(UpdaterContext& context) const
+    {
+        // Take the 'url' as the input file path.
+        auto url {context.spUpdaterBaseContext->configData.at("url").get_ref<const std::string&>()};
+        constexpr auto filePrefix {"file://"};
+        if (Utils::startsWith(url, filePrefix))
+        {
+            Utils::replaceFirst(url, filePrefix, "");
+        }
+        const std::filesystem::path inputFilePath {url};
+
+        // Check if file is compressed.
+        const auto compressed {
+            "raw" != context.spUpdaterBaseContext->configData.at("compressionType").get_ref<const std::string&>()};
+
+        // Generate output file path. If the input file is compressed, the output file will be in the downloads folder
+        // and if it's not compressed, in the contents folder.
+        auto outputFilePath {compressed ? context.spUpdaterBaseContext->downloadsFolder
+                                        : context.spUpdaterBaseContext->contentsFolder};
+        outputFilePath = outputFilePath / inputFilePath.filename();
```

Comment:
All of this is only used inside the if, on line 114. Can it be moved there?

Reply:
Aah missed that. Fixed.